state-chain/runtime/src/chainflip/bitcoin_elections.rs

```diff
@@ -299,56 +290,279 @@ impl Hook<btc::BlockNumber, Vec<DepositChannelDetails<Runtime, BitcoinInstance>>
-		} else {
-			log::info!("Processing witnesses: {:?} for block {:?}", witnesses, current_block);
+	fn default() -> Self {
+		Self {
+			blocks_data: Default::default(),
+			reorg_events: Default::default(),
+			rules: {
+				let mut result = BoundedVec::<BtcRuleType, ConstU32<2>>::new();
+				result.try_push(BtcRuleType::PreWitness).expect("cannot fail");
+				result.try_push(BtcRuleType::Witness).expect("cannot fail");
+				result
+			}
 		}
-		for (deposit_block_number, deposits) in witnesses.clone() {
-			for deposit in deposits {
-				if deposit_block_number == current_block {
-					log::info!("Prewitness deposit submitted by election: {:?}", deposit);
-					let _ = BitcoinIngressEgress::process_channel_deposit_prewitness(
-						deposit,
-						deposit_block_number,
-					);
-				} else if let Some(safety_margin) =
-					WitnessSafetyMargin::<Runtime, BitcoinInstance>::get()
-				{
-					if deposit_block_number <= (current_block - safety_margin) {
-						log::info!("deposit election submitted by election: {:?}", deposit);
-						BitcoinIngressEgress::process_channel_deposit_full_witness(
-							deposit,
-							deposit_block_number,
-						);
+	}
+}
+
+impl BlockWitnesserProcessor<BlockNumber, BlockData, BtcEvent>
+	for DepositChannelWitessingProcessor<BlockNumber, BlockData, BtcEvent>
+{
+	fn process_block_data(&mut self, chain_progress: ChainProgressInner<BlockNumber>) {
+		match chain_progress {
+			ChainProgressInner::Progress(last_height) => {
+				let last_events = deduplicate_btc_events(self.process_rules(last_height));
+				execute_events(last_events);
+				self.clean_old(last_height);
+			},
+			ChainProgressInner::Reorg(range) => {
+				for n in range {
+					let block_data = self.blocks_data.remove(&n);
+					if let Some((data, last_age)) = block_data {
+						// We need to get only events already processed
+						for age in 0..last_age {
+							self.reorg_events
+								.insert(n, self.process_rules_for_age_and_block(n, age, &data));
+						}
 					}
 				}
+			},
+		}
+	}
+
+	fn insert(&mut self, n: BlockNumber, block_data: BlockData) {
+		self.blocks_data.insert(n, (block_data, 0));
+	}
+
+	fn clean_old(&mut self, n: BlockNumber) {
+		self.blocks_data.retain(|key, (data, age)| *age <= BitcoinIngressEgress::witness_safety_margin().unwrap());
+		self.reorg_events.retain(|key, data| *key > n - BUFFER_EVENTS);
+	}
+
+	fn process_rules(&mut self, last_height: BlockNumber) -> Vec<BtcEvent> {
+		warn!("Processing rules... last_height: {last_height:#?}");
+		let mut last_events: Vec<BtcEvent> = vec![];
+		for (block, (data, last_age)) in self.blocks_data.clone() {
+			warn!("Rules for block {block:?}, age: {last_age:?}, data: {data:?}");
+			for age in last_age..=last_height - block {
+				last_events = last_events
+					.into_iter()
+					.chain(self.process_rules_for_age_and_block(block, age, &data))
+					.collect();
 			}
+			self.blocks_data.insert(block, (data.clone(), last_height - block + 1));
```

Comment:
I might be misreading sth here, but are we reinserting into the `blocks_data` array which already contains this `block` and `data`? So the same data is going to appear multiple times if it was processed multiple times?